File: bench/hashmap.c
/* Hash-map benchmark: open-addressing (linear-probe) table over colliding
   integer keys, last write wins, then sum looked-up values.
   Matches hashmap.zeph/.rs (same keys, same last-write-wins semantics). */
#include <stdio.h>
#include <stdlib.h>

#define CAP (1 << 21)   /* 2,097,152 slots, comfortably > 500,009 live keys */

int main(int argc, char **argv) {
    long N = argc > 1 ? atol(argv[1]) : 2000000;
    long long *keys = malloc((size_t)CAP * sizeof(long long));
    long long *vals = malloc((size_t)CAP * sizeof(long long));
    char *used = calloc(CAP, 1);

    for (long i = 0; i < N; i++) {
        long long k = ((long long)i * 2654435761LL) % 500009;
        unsigned long long h = (unsigned long long)k * 11400714819323198485ULL;
        long slot = (long)(h >> 43) & (CAP - 1);
        while (used[slot] && keys[slot] != k) slot = (slot + 1) & (CAP - 1);
        used[slot] = 1; keys[slot] = k; vals[slot] = i;
    }

    long long chk = 0;
    for (long i = 0; i < N; i++) {
        long long k = ((long long)i * 2654435761LL) % 500009;
        unsigned long long h = (unsigned long long)k * 11400714819323198485ULL;
        long slot = (long)(h >> 43) & (CAP - 1);
        while (used[slot] && keys[slot] != k) slot = (slot + 1) & (CAP - 1);
        chk += used[slot] ? vals[slot] : 0;
    }
    printf("%lld\n", chk);
    return 0;
}

File: bench/hashmap.rs
// Hash-map benchmark: std HashMap over colliding integer keys, last write wins,
// then sum looked-up values. Matches hashmap.zeph/.c.
use std::collections::HashMap;
use std::env;

fn main() {
    let a: Vec<String> = env::args().collect();
    let n: i64 = if a.len() > 1 { a[1].parse().unwrap() } else { 2000000 };
    let mut m: HashMap<i64, i64> = HashMap::new();
    for i in 0..n {
        m.insert((i * 2654435761) % 500009, i);
    }
    let mut chk: i64 = 0;
    for i in 0..n {
        chk += *m.get(&((i * 2654435761) % 500009)).unwrap_or(&0);
    }
    println!("{}", chk);
}
